generic/fdstat_sysctl.c

```diff
@@ -0,0 +1,85 @@
+/*
+htop - generic/fdstat_sysctl.c
+(C) 2022-2023 htop dev team
+Released under the GNU GPLv2+, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "generic/fdstat_sysctl.h"
+
+#include <math.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <sys/sysctl.h>
+#include <sys/types.h>
+
+#include "config.h"
+
+
+static void Generic_getFileDescriptors_sysctl_internal(
+   const char* sysctlname_maxfiles,
+   const char* sysctlname_numfiles,
+   size_t size_header,
+   size_t size_entry,
+   double* used,
+   double* max
+) {
+   *used = NAN;
+   *max = 65536;
+
+   int max_fd, open_fd;
+   size_t len;
+
+   len = sizeof(max_fd);
+   if (sysctlname_maxfiles && sysctlbyname(sysctlname_maxfiles, &max_fd, &len, NULL, 0) == 0) {
+      if (!max_fd) {
+         max_fd = 1;
+      }
+      *max = max_fd;
+   }
+
+   len = sizeof(open_fd);
+   if (sysctlname_numfiles && sysctlbyname(sysctlname_numfiles, &open_fd, &len, NULL, 0) == 0) {
+      *used = open_fd;
+   }
+
+   if (!isnan(*used)) {
+      return;
+   }
+
+   // If no sysctl arc available, try to guess from the file table size at kern.file
+   // The size per entry differs per OS, thus skip if we don't know:
+   if (!size_entry)
+      return;
+
+   len = 0;
+   if (sysctlbyname("kern.file", NULL, &len, NULL, 0) < 0)
+      return;
+
+   if (len < size_header)
+      return;
+
+   *used = (len - size_header) / size_entry;
+}
+
+void Generic_getFileDescriptors_sysctl(double* used, double* max) {
+#if defined(HTOP_DARWIN)
+   Generic_getFileDescriptors_sysctl_internal(
+      "kern.maxfiles", "kern.num_files", 0, 0, used, max);
+#elif defined(HTOP_DRAGONFLY)
+   Generic_getFileDescriptors_sysctl_internal(
+      "kern.maxfiles", NULL, 0, sizeof(struct kinfo_file), used, max);
+#elif defined(HTOP_FREEBSD)
+   Generic_getFileDescriptors_sysctl_internal(
+      "kern.maxfiles", "kern.openfiles", 0, sizeof(struct xfile), used, max);
+#elif defined(HTOP_NETBSD)
+   Generic_getFileDescriptors_sysctl_internal(
+      "kern.maxfiles", NULL, sizeof(struct filelist), sizeof(struct file), used, max);
```

Comment:
```suggestion
      " "kern.maxfiles", NULL, 0, sizeof(struct kinfo_file), used, max);
```

This change made it compile in NetBSD, not sure where it is getting the `164` from.

![image](https://user-images.githubusercontent.com/189939/216846738-d3793e63-524d-41aa-9655-f8aee6f9aed4.png)

Reply:
The internal function has two modes:
1. Get the number of open files directly from the sysctl by the name given in the second parameter.
2. Get the number of open files indirectly by retrieving the information about open files via sysctl `kern.files`, subtracting a fixed header (the `0` in your case) and dividing the remainder by the size per structure (`size(struct kinfo_file)`).

Thus in your case the 164 resulted from the second approach.

FWIW: The max value is read from the sysctl given in the first argument.